madmom/ml/hmm.pyx

```diff
@@ -518,52 +550,61 @@ class HiddenMarkovModel(object):
         """
         # transition model stuff
         tm = self.transition_model
-        cdef unsigned int [::1] tm_states = tm.states
-        cdef unsigned int [::1] tm_pointers = tm.pointers
+        cdef uint32_t [::1] tm_states = tm.states
+        cdef uint32_t [::1] tm_pointers = tm.pointers
         cdef double [::1] tm_probabilities = tm.probabilities
-        cdef unsigned int num_states = tm.num_states
+        cdef int num_states = tm.num_states
 
         # observation model stuff
         om = self.observation_model
-        cdef unsigned int num_observations = len(observations)
-        cdef unsigned int [::1] om_pointers = om.pointers
-        cdef double [:, ::1] om_densities = om.densities(observations)
+        cdef uint32_t [::1] om_pointers = om.pointers
+        # TODO: check why ascontiguousarray is needed for framewise processing
+        # make sure we can iterate over the observations
+        cdef double [:, ::1] om_densities = np.ascontiguousarray(
+            np.atleast_2d(om.densities(observations)))
+        cdef int num_observations = len(om_densities)
+
+        # reset HMM
+        if reset:
+            self.reset()
 
         # forward variables
-        cdef double[:, ::1] fwd = np.zeros((num_observations + 1, num_states),
-                                           dtype=np.float)
+        cdef double[::1] fwd_prev = self._prev
+        cdef double[:, ::1] fwd_cur = \
+            np.zeros((num_observations, num_states), dtype=np.float)
```

Comment:
1. ok, this makes sense.
2. `fwd_cur` holds only `num_observations`, not `num_observations + 1` variables any more. The previous variables are stored in `self._prev`. This change was necessary (at least I think it is) because during initialisation we know how many states we have and can thus init `self._prev`, but don't know the lengths of the observations yet. If we want to be able to call `forward()` with observations of variable length (i.e. both a complete sequence or frame-by-frame), we have to init `fwd_cur` (or `fwd`, see 1.) inside `forward()` but need to keep the previous variables somewhere, hence `self._prev`.